src/osdnet.c

```diff
@@ -1432,6 +1444,96 @@ osn_pfdeinit(struct pfdata *pfdata, struct osnpf *osnpf)
 	pfdata->pf_deinit(pfdata, osnpf);
 }
 
+#if KLH10_NET_LNX
+
+/*
+  The Linux PF_PACKET interface is described to some extent
+  by the packet(7) man page.
+
+  Linux provides no kernel packet filtering mechanism other than
+  possibly a check on the ethernet protocol type, but this is useless
+  for us since we'll always want to check for more than just one type;
+  e.g. IP and ARP, plus possibly 802.3 or DECNET packets.
+
+  From the man page for packet(7):
+       By  default all packets of the specified protocol type are
+       passed to a packet socket. To only get packets from a spe-
+       cific  interface  use  bind(2)  specifying an address in a
+       struct sockaddr_ll to bind the packet socket to an  inter-
+       face.  Only  the  sll_protocol and the sll_ifindex address
+       fields are used for purposes of binding.
+ */
+void
+osn_pfinit_lnx(struct pfdata *pfdata, struct osnpf *osnpf, void *pfarg)
+{
+    int fd;
+    char *ifcname = osnpf->osnpf_ifnam;
+    struct ifreq ifr;
+
+    /* Open a socket of the desired type.
+     */
+    struct sockaddr_ll sll;
+    int ifx;
+
+    /* Get raw packets with ethernet headers
+     */
+    fd = socket(PF_PACKET, SOCK_RAW,
+#if 0 /*OSN_USE_IPONLY*/		/* If DPIMP or otherwise IP only */
+		htons(ETH_P_IP)		/* for IP only */
+#else
+		htons(ETH_P_ALL)	/* for everything */
+#endif
+		);
+    if (fd < 0)
+	esfatal(1, "Couldn't open packet socket");
+
+    if (!osnpf->osnpf_dedic) {
+      esfatal(1,"LNX ifmethod only works for dedicated interfaces");
+    }
+
+    /* Need ifc index in order to do binding, so get it. */
+    memset(&ifr, 0, sizeof(ifr));
+    strncpy(ifr.ifr_name, ifcname, sizeof(ifr.ifr_name));
+    if (ioctl(fd, SIOCGIFINDEX, &ifr) < 0 )
+	esfatal(1, "SIOCGIFINDEX of %s failed", ifcname);
+    ifx = ifr.ifr_ifindex;
+
+    /* Bind to proper device/interface using ifc index */
+    memset(&sll, 0, sizeof(sll));
+    sll.sll_family = AF_PACKET;
+    sll.sll_protocol = htons(ETH_P_ALL);
+    sll.sll_ifindex = ifx;
+    if (bind(fd, (struct sockaddr *)&sll, sizeof(sll)))
+	esfatal(1, "bind to %s failed", ifcname);
+
+    /* This code only works with Ethernet, so check for that */
+    memset(&ifr, 0, sizeof(ifr));
+    strncpy(ifr.ifr_name, ifcname, sizeof(ifr.ifr_name));
+    if (ioctl(fd, SIOCGIFHWADDR, &ifr) < 0 )
+	esfatal(1, "SIOCGIFHWADDR of %s failed", ifcname);
+
+    if (ifr.ifr_hwaddr.sa_family != ARPHRD_ETHER)
+	efatal(1, "%s is not an ethernet - ARPHRD type %d",
+	       ifcname, ifr.ifr_hwaddr.sa_family);
+
+    /* Finally, attempt to determine current ethernet MAC address.
+       Assume above call returned it in sa_data.
+     */
+    ea_set(&osnpf->osnpf_ea, &ifr.ifr_addr.sa_data[0]);
+
+    pfdata->pf_meth = PF_METH_LNX;
+    pfdata->pf_read = osn_pfread_fd;
+    pfdata->pf_write = osn_pfwrite_fd;
```

Comment:
Small nit: The definition of `osn_pfread_fd` and `osn_pfwrite_fd` are `#ifdef`ed with a different condition. When configuring TUN and TAP out it will be missing. Easy to fix of course.